### ISSUE TITLE
False error for missing positional argument in TypeGuard.

### ISSUE THREAD

**Issue by DolajoCZ** [python/mypy#15090]
**Bug Report**

False error for missing positional argument in `TypeGuard`.

**To Reproduce**

```python
from __future__ import annotations
from typing import TypeGuard, Generic, TypeVar, Any

import dataclasses


T = TypeVar("T")


@dataclasses.dataclass(frozen=True, eq=True)
class Foo(Generic[T]):

    data: T

    def is_str(self) -> TypeGuard[Foo[str]]:
        return isinstance(self.data, str)

    def evaluate(self) -> None:
        if self.is_str():
            print(self.data)
```

**Expected Behavior**

Without any error.

**Actual Behavior**

```
main.py:15: error: TypeGuard functions must have a positional argument  [valid-type]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.2.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files): `strict`
- Python version used: 3.10.11

**Comment by erictraut** [comment]
This error is correct. You need to provide at least one parameter (beyond the `self` parameter) for a user-defined type guard function. For details, refer to [PEP 647](https://peps.python.org/pep-0647). Here's the relevant statement:

> If a type guard function is implemented as an instance method or class method, the first positional argument maps to the second parameter (after “self” or “cls”).

And [here](https://peps.python.org/pep-0647/#narrowing-of-implicit-self-and-cls-parameters) is the reasoning behind this decision.
